src/PkgAuthentication.jl

```diff
@@ -171,23 +171,117 @@ to RequestLogin, or to Failure otherwise.
+# device authentication endpoints. Returns a Dict with the following
+# fields:
+# - `auth_flows`::Vector{String}: The authentication mechanisms supported
+#   by the server. Eg: ["classic", "device"]
+# - `device_token_refresh_url`::String: The refresh URL for refreshing the auth
+#   token
+# - `device_authorization_endpoint`::String: The endpoint that must
+#   be called to initiate device flow authentication. This field is
+#   only present when device flow is enabled on the server.
+# - `device_token_endpoint`::String: The endpoint that should be called to
+#   retrieve the authentication token after the user has approved
+#   the authorization request. This field is only present when device
+#   flow is enabled on the server.
+function get_auth_configuration(state::NoAuthentication)
+    output = IOBuffer()
+    auth_suffix = isempty(state.auth_suffix) ? "auth" : state.auth_suffix
+    response = Downloads.request(
+        "$(state.server)/$(auth_suffix)/configuration",
+        method = "GET",
+        output = output,
+        throw = false,
+        headers = ["Accept" => "application/json"],
+    )
+
+    if response isa Downloads.Response && response.status == 200
+        body = nothing
+        content = String(take!(output))
+        try
+            body = JSON.parse(content)
+        catch ex
+            @debug "Request for well known configuration returned: ", content
+            return Dict{String, Any}()
+        end
+
+        if body !== nothing
+	    @assert !haskey(body, "auth_flows") || !("device" in body["auth_flows"]) || (haskey(body, "device_authorization_endpoint") && haskey(body, "device_token_endpoint") && haskey(body, "device_token_refresh_url"))
+            return body
+        end
+    end
+
+    return Dict{String, Any}()
 end
-Base.show(io::IO, s::NoAuthentication) = print(io, "NoAuthentication($(s.server))")
 
 function step(state::NoAuthentication)::Union{RequestLogin, Failure}
-    challenge = Random.randstring(32)
+    auth_config = get_auth_configuration(state)
+    success, challenge, body_or_response = if "device" in get(auth_config, "auth_flows", [])
+        fetch_device_code(state, auth_config["device_authorization_endpoint"])
+    else
+        initiate_browser_challenge(state)
+    end
+    if success
+        return RequestLogin(state.server, state.auth_suffix, challenge, body_or_response, get(auth_config, "device_token_endpoint", ""), get(auth_config, "device_token_refresh_url", ""))
+    else
+        return HttpError(body_or_response)
+    end
+end
+
+function fetch_device_code(state::NoAuthentication, device_endpoint::AbstractString)
     output = IOBuffer()
     response = Downloads.request(
-        string(state.server, "/challenge"),
+        device_endpoint,
+        method = "POST",
+        input = IOBuffer("client_id=$(get(ENV, "JULIA_PKG_AUTHENTICATION_DEVICE_CLIENT_ID", "device"))&scope=openid email profile offline_access"),
```

Comment:
@nkottary @tanmaykm One final concern I have here is the `scope=` value. The OAuth2 standard doesn't specify what the value should be here (in fact, it's optional). The specific scopes (`openid email profile offline_access`) seem very Dex specific? So this wouldn't work with a different IdP?

I think we need to add that to the configuration endpoint to. E.g. optional `device_token_scope` value, which defines the scope PkgAuthentication SHOULD request (if present). The scope from the server MUST minimally provide package server access (but can be broader).

Another option would be to maybe omit scope by default, but I don't think Dex allows that?